### ISSUE TITLE
codegen: Chaining syntax - When to execute query (i.e., implicit vs explicit)?

### ISSUE THREAD

**Issue by helderco** [dagger/dagger#3555]
I started thinking about this for Python but we got feedback from user _Luis Guzman_ on Discord: [Chaining syntax, Confusing behavior](https://discord.com/channels/707636530424053791/1034734239675068458).

## Go SDK

> [...] I've found the chaining syntax a bit... unintuitive.
> 
> Initially, it's just not very clear what actually executes the GQL, but ultimately I've figured out - just find whatever takes a context. 

Even then the user gives an example that he found confusing. I'll try it next to understand but for now...

## Python SDK

In Python it's worse, because it's simpler:

```python
alpine = api.container().from_("python:3.10.8-alpine")
version = alpine.exec(["python", "-V"]).stdout().contents()
assert version == "Python 3.10.8\n"
```

The query executes automatically when you call a method that returns a leaf value (i.e., scalar or enum) but this implicit behavior may be too confusing.

I'm wondering what's better:
```python
# Pass client session as argument to leaf as in Go
...stdout().contents(session)

# Execute explicitly:
value = session.execute(version)

# Return leaf as an object with `execute`
...stdout().contents().execute()
```

The last two options are also more flexible, opening up the possibility of selecting sibling fields into a single query.

## Prior art

Database queries in Django return lazy objects. They are resolved when they're used. It's not exactly explicit but it's **documented** that some [methods return a new QuerySet](https://docs.djangoproject.com/en/4.1/ref/models/querysets/#methods-that-return-new-querysets), while [others return results](https://docs.djangoproject.com/en/4.1/ref/models/querysets/#methods-that-do-not-return-querysets).

It's also documented [when the query is evaluated](https://docs.djangoproject.com/en/4.1/ref/models/querysets/#when-querysets-are-evaluated). It's not intuitive if you don't read the documentation otherwise you could be making a bunch of unnecessary queries. We're not optimizing for that yet, what I'm focusing on is the implicit vs explicit behavior.

Is documentation enough to keep usage simple or should we prefer more verbose but also more explicit usage?

**Comment by sipsma** [comment]
Just some initial unsorted thoughts. 

---

I think the options you listed are all worth considering, I do have some questions about the current state too though.

In this example:
```python
version = alpine.exec(["python", "-V"]).stdout().contents()
assert version == "Python 3.10.8\n"
```

If you had skipped the `.contents()` then version would be some sort of object, not a string, right? And so wouldn't IDEs and the mypy linter (or however that works in python) notice that and highlight it/fail the type check?

I guess that's sort of the current approach in general: if the returned type is either a primitive value or a list/object consisting just of primitive values, then that means evaluation will run.

I'm not saying that's better than the other options, just kind of thinking out loud and trying to verify assumptions about type checking in Python are correct.

---

Luis Guzman on discord also said that the explicit execute makes sense:
```
makes some sense to me with an explicit call to eval. Especially considering your earlier comments about multiple commands being put together for more efficient llb solves. I could imagine chaining multiple commands together, then making a bigger call to execute() , with LLB doing the dirty work of solving parallel execution (like multi-stage docker)
```

I think we discussed something like that way early on but figured it would be unintuitive; was nicer to try and hide the laziness. Maybe it's worth reconsidering that though?

--- 

> Is documentation enough to keep usage simple

I guess this is the absolute bare minimum we have to do; doc strings on the objects should indicate whether they are synchronously evaluated or not.

---

As a one off case, I think we should even consider just making `Stdout` and `Stderr` return strings directly rather than requiring an additional `.Contents`. That doesn't solve the general problem but it feels more intuitive to me.

**Comment by helderco** [comment]
> [...] wouldn't IDEs and the mypy linter (or however that works in python) notice that and highlight it/fail the type check?

It depends on the usage. In this example there'd be no warning because we're doing a comparison. It'll just return `False` that `File` isn't that string.

If you try to access a string attribute (e.g., `.lower()`), in VS Code there's no warning:

<img width="393" alt="Screen Shot 2022-10-27 at 13 15 41" src="https://user-images.githubusercontent.com/174525/198294634-0e93a5db-e9f0-43dd-9d24-d76214f7d5b2.png">

I may need to tune _Pyright_ (VS Code's type checker from Microsoft), not sure what's happening there. For now I've been focused on `mypy` which does raise the issue:

<img width="484" alt="Screen Shot 2022-10-27 at 13 11 07" src="https://user-images.githubusercontent.com/174525/198294125-71228df4-681a-48da-b778-b2f825a8801c.png">

**Comment by helderco** [comment]
For the "Return leaf as an object with `execute`" case I made a quick experiment:

<img width="483" alt="Screen Shot 2022-10-27 at 12 56 29" src="https://user-images.githubusercontent.com/174525/198296293-89c8b36c-f078-4945-9d8f-afec6f223df8.png">

So this tells you that the `version` variable is a `Result` with a `str` value and we can say that these are executed when you try to access that `.value`. It's both lazy but also conventional.

You could also use it to call execute manually:

```python
try:
    value = version.execute()  # inferred `value: str` because `version: Result[str]`
except Error:
    ....
```

I ok with the lazy behavior since I'm used to it. In Django you use laziness everywhere.

**Comment by helderco** [comment]
My opinion is that I'm ok with a some "implicit" if it has a convention. Either if it's passing the `ctx` in Go or just knowing that leaf values cause execution. Problem with that is that if it's a custom scalar it won't be obvious that's a leaf value. I can fix that with the `Result` or `Leaf` wrapper object.

In Go, you can also tell you're executing if you need to capture an error. I could do the same in Python, problem is it's not very Pythonic:

```python
version, err = stdout.Contents()

# vs

try:
    value = stdout.Contents()
except Error as err:
    ...
```

I don't want to be blocked on this so I think I'll move ahead with returning the wrapper object. It should be ok to change later if faced with overwhelming feedback in favor of something a bit more explicit like:

```python
try:
    version = stdout.Contents().execute(session)
except Error as err:
    ....
```

**Comment by jlongtine** [comment]
I think this (or something similar) is biting @plaisted https://github.com/dagger/dagger/issues/3551

**Comment by shykes** [comment]
One observation, one suggestion, and one follow-up question.

**Observation**: we are discussing this as a general issue that applies to all core. But in reality 100% of the problems experienced are related to `exec`. Basically "when does my exec actually exec, and when does it not". Given that, I think we can add to the table a bunch of solutions that are specific to exec, instead of only looking for hypothetical generalized solutions.

**Suggestion**: we make `exec` non-lazy by default, with an optional `lazy` option for power users. As @aluzzardi explained to me yesterday, buildkit does not have such an option that we can use cleanly. But I don't think we need it: we just implement the lazy/non-lazy behavior at the graphql resolver layer. If non-lazy: always ask buildkit for, say, the exitcode (I know it's indirectly via the shim etc, but you get the idea: query buildkit in such a way that it actually executes).

**Question**: assuming that the above suggestion is technically feasible, which of the problems described in this issue would still remain? My instinctive answer is: little to none. But I may be wrong.

Thanks for reading.

**Comment by aluzzardi** [comment]
**Summary of the situation**

Today, some functions are "lazy" and don't result in an execution (e.g. `Stdout()`) while others execute  (e.g. `Stdout().Contents(ctx)`).

The model is implicit (depending on whether it's a leaf node or not) with a "rule of thumb" in each language to hint which one is which:
- **Go**: functions taking a context
- **Python**: async functions
- **Node**: functions returning a Promise

**The problems with this are**

- **This model is confusing as reported in multiple instances**
  - Have to understand the "rule of thumb" in each language
  - Hides the core laziness concept behind magic
- **Python** (@helderco: correct me if I'm wrong)
  - "rule of thumb" is not as clear as Go (no context and no error return)
  - "rule of thumb" is the async keyword: only works with async Python, no equivalent for sync
  - Python sync/async **may** require one function per resolver (e.g. `contents()` vs `syncContents()`)
- **Node**
  - @slumbering mentioned implicit doesn't work in TypeScript

**Proposal: Make execution explicit**

**Possible Solutions**
- **Explicit Inlined Execution**
  - Add a method (for instance, `Execute()`) at the end of chaining to trigger execution
  - Example: `out, err := ...Stdout().Contents().Execute(ctx)`
- **Separate Execution Phase**
  - Build the query and then execute it using a special top-level client function
    - `q := ...Exce().Foo().Bar(); client.Query(q)`
  - More context here: https://github.com/dagger/dagger/issues/3558#issuecomment-1297655000

**Comment by aluzzardi** [comment]
> The rule of thumb is actually getting a Result object

Right. What I meant is, in Go, if you don't follow the rule of thumb, the code will NOT compile. Missing context, wrong number of return arguments. In Python it's much more subtle: you have to look at the return type from the editor hints.

**Comment by aluzzardi** [comment]
@shykes some raw notes

> **Suggestion**: we make exec non-lazy by default, with an optional lazy option for power users
> **Question**: assuming that the above suggestion is technically feasible, which of the problems described in this issue would still remain

- The same problems would remain: even if exec is non-lazy, stdout is still implicit and non-lazy
- The problem is generalized
  - exec is when people realize there's a problem: they're expecting the code to take some time to run, or to see their output printed in the logs. But they encountered the same problem before without even realizing
  - `Git().Tree()` doesn't actually clone
  - `Host().Directory()` doesn't copy files
  - `WithCopiedFile` doesn't copy the file
  - It's just that those operations are fast and the user has no log output expectation, so they're not realizing yet the confusion
  - IMHO, special casing Exec "postpones" the confusion at a later stage, and it's slightly amplified because "it worked with exec"
- Making exec non-lazy means we break buildkit parallelism
  - Take @sipsma's multi-arch demo: if `exec` were non-lazy, then the `for` loop building each platform would need to be rendered asynchronous using goroutines / async python / something in node
- Kills a bunch of optimizations
  - alpine example: the for loop installing packages would do a back and forth with buildkit for each package, rather than sending in batch
  - I think it's swimming against the buildkit flow
- DX suffers: undo a lot of improvements made possible by "hide IDs"
  - It would be impossible to chain Execs (like our magefile is doing)
  - Becomes less composable

**Comment by shykes** [comment]
> @shykes some raw notes
> [...] bad idea [...]

I just realized my suggestion was mixed with another implicit suggestion, which is to better separate the "pipeline builder". So, let me withdraw my previous comment and try again, focusing on the pipeline builder part first.

**Comment by aluzzardi** [comment]
Elaborating the **Explicit Inlined Execution** which to me right now is the one that makes most sense.

- Nothing is ever implicit -- the API is always lazy (you're always building a "query")
- There is an explicit `Execute()`, across all languages
- optional nit: `Exec()` is renamed `Command()` or something

```go
// lazy
container := c.Container().From("alpine").Command([]string{"echo", "foo"})

// Optional: Explicitly force execution of container (somehow -- not sure how it'd work) whenever you want
container, err = container.Execute(ctx)

// Explicitly execute to grab a field
stdout, err := container.Stdout().Contents().Execute(ctx)
```

Rule:
- **Execute** executes the query, including its parents and siblings (dependencies)
- If you didn't `Execute`, it didn't happen
- `Execute` is cached -- `Executing` the same nodes will yield the same result and is a no op if already done

**Comment by shykes** [comment]
> **Proposal: Make execution explicit**
> 
> **Possible Solutions**
> [...]
>   * Build the query and then execute it using a special top-level client function
>     
>     * `q := ...Exec().Foo().Bar(); client.Query(q)`
>   * More context here: [Consider hiding the need for IDs in codegen clients  #3558 (comment)](https://github.com/dagger/dagger/issues/3558#issuecomment-1297655000)

As you know, this is my preferred solution :) I call it the "**pipeline builder**" model. There are unresolved questions on implementation and DX, but fundamentally I think this is the simplest model to explain and understand. IMO until we have this clear separation, our model will be awkward to explain and understand (even if the implementation is solid and the DX "makes sense once you get used to it" -> much better if the model is so natural that you don't need to "get used to it": it just intuitively makes sense).

So far I have listed 3 DX options to implement "pipeline builder" in Go. NOTE: open question on how well these translate to Python and NodeJS.

1. Pointers
2. Callbacks
3. Collapsed output object

## 1. Pointers

In the pointer approach, you insert "getFoo" operations in the pipeline, with a pointer to the destination value. Later, when the pipeline is executed, the destination value is filled.

Example:

```go
var (
  lsStdout string
  lsCode int
)
pipeline := dagger.Container().
  From("alpine").
  Exec([]string{"ls"}).
  GetExitCode(&lsCode).
  GetStdout(&lsStdout).
  Exec([]string{"echo", "something", "else"})

err := dagger.Connect(ctx).Query(pipeline)
```

## 2. Callbacks

In this approach, you also insert a "getFoo" operation in the pipeline, but you pass a callback instead of a pointer. This is more flexible - the pointer option can be implemented on top of the callback option, but not the other way around.

```go
var (
  lsStdout string
  lsCode int
)
pipeline := dagger.Container().
  From("alpine").
  Exec([]string{"ls"}).
  GetExitCode(func(code int) error { lsStdout = code; return nil } ).
  GetStdout(func(stdout string) error { lsStdout = stdout; return nil } ).
  Exec([]string{"echo", "something", "else"})

err := dagger.Connect(ctx).Query(pipeline)
```

## 3. Collapsed output object

In this approach, you insert a "getFoo" operation, but unlike the pointer and callback options, you *do not pass an argument*. Instead, pipeline execution does pretty much what the underlying graphql layer does: return an object with all the values combined. There is one twist: in order to avoid tight coupling to the underlying graphql query, the output object would be *collapsed*.

Collapsing could be done either automatically, or manually.

### 3a. Auto-collapsed output object

```go
pipeline := dagger.Container().
  From("alpine").
  Exec([]string{"ls"}).
  GetExitCode().
  GetStdout().
  Exec([]string{"echo", "something", "else"}).
  GetExitCode()

o, _ := dagger.Connect(ctx).Query(pipeline)
otherExitCode, _ := o.GetInt("exitCode") // last one wins
// fully qualified alternative:
// otherExitCode, _ = o.GetInt("container.from.exec.exec.exitCode")

lsExitCode, _ := o.GetInt("container.from.exec.exitCode")

lsStdout, _ := o.GetString("stdout")
// fully qualified alternative:
// lsStdout, _ = o.GetString("container.from.exec.stdout")
```

### 3b. Manually collapsed output object

```go
pipeline := dagger.Container().
  From("alpine").
  Exec([]string{"ls"}).
  GetExitCode("lsExitCode").
  GetStdout("lsStdout").
  Exec([]string{"echo", "something", "else"}).
  GetExitCode("otherExitCode")

o, _ := dagger.Connect(ctx).Query(pipeline)
otherExitCode, _ := o.GetInt("otherExitCode")
lsExitCode, _ := o.GetInt("lsExitCode")
lsStdout, _ := o.GetString("lsStdout")
```

## Preference

At the moment my preference is either option 2 (callbacks) or 3b (manually collapsed output object).

|   | Pros | Cons |
| -- | -- | -- |
| Callbacks | Flexible; familiar to Go devs; no type casts | More boilerplate for simple things; may not translate to all languages |
| Manually collapsed output object | Simpler pipeline declaration; less overhead for simple cases; maps closely to the underlying graphql model | Type casts; less familiar model, need to understand what the "key" argument is for |

**Comment by samalba** [comment]
I see how a long chained query will produce a better dag behind, it's declarative, we get clarity on what the backend will run, it's optimized, etc...

However as a dev, I very much prefer very small queries like we have in the tutorial today:

```
myRepo := client.Git(repoURL)
myBaseImage := golang := client.Container().From("golang:latest")
resultImage = myBaseImage.Exec(..)
```

I don't have to understand that calling `Exec()` later will actually resolve the pipeline, which removes a lot of cognitive complexity while I write the pipeline. For example, I can call another API while building the pipeline.

True, I can easily end up with a pipeline that is not as optimized as if the SDK was forcing me to chain everything at once. But I prefer it this way, it also makes the code is more readable for someone who gets into it and does not dagger.

**Comment by vito** [comment]
Speaking only on the Go SDK: I agree with @samalba - I think we're closer to the expressiveness sweet spot right now, and this optimization takes a step into the point of diminishing returns.

I think it should be OK for Dagger code to look and behave imperative even if it's really declarative under the hood. It leverages the caching semantics to the point where users can write more intuitive code with only a slight performance penalty in the grand scheme of things (construct quer + send request + build LLB => cache hit).

For other SDKs: I really think the answer is "it depends." Maybe the pipeline/builder pattern is appropriate in some languages that don't have the `ctx`/`err` heuristic. In Haskell for example this would make a ton of sense, since I would want to avoid the `IO` monad as much as possible.[^1] So to me the decision on whether a SDK should have builder UX or imperative UX or [...] should be mostly influenced by the language, and we don't need it to be consistent across them all.

Re: user confusion - I think this is something we should keep an eye but try not to overreact to. If the majority of users are complaining, sure, but if it's not many people and they figure it out pretty quickly and never think about it again I think that's OK. There could also be many reasons for this. Maybe it's just a naming problem; `Exec` sounds more synchronous than `WithX`. It's one of only a few verbs and it has a `FIXME` on it still. :sweat_smile: 

[^1]: For those who don't know Haskell, IO essentially a jail/virus - once a function needs to use the IO monad, _every_ function that calls it needs to as well, and your code starts to appear more imperative.

**Comment by aluzzardi** [comment]
@samalba 
> But I prefer it this way, it also makes the code is more readable for someone who gets into it and does not dagger.

@vito 
> Speaking only on the Go SDK: I agree with @samalba - I think we're closer to the expressiveness sweet spot right now, and this optimization takes a step into the point of diminishing returns.

@helderco 
You didn't say this explicitly, but I see you implemented the implicit approach -- I assume you prefer it as well?

@sipsma 
Thoughts?


I agree with the simplicity and familiarity statements -- perhaps it's just something we have to document and give some time for folks to get familiar with. If everyone feels that way, I'm ok with continuing the current path, see what feedback comes back, and adjust (or not).

**Comment by TomChv** [comment]
I also agree with @samalba, I fear that the DX can become verbose and make the code too complex for simple things.
I was playing with Cloak this morning and I feel really comfortable with the current shape of the Go SDK.

However I'm concerned about the Typescript SDK, since the execution needs to be `async`, we cannot really apply for an implicit pattern.
The problem is that if we have implicit Go and Python SDK but explicit Typescript SDK, we'll have a design conflict that may lead to further problems later. (How do we implement multi-platform in all those SDK for instance).

I do not have the absolute answer to the question but I'm concerned if we end up with differents approaches across SDK

**Comment by shykes** [comment]
I agree there's a risk of going too far in the chaining model, and the current status quo may be the perfect balance.

But, to play devil's advocate:

* If it's the perfect balance, why are we having this conversation about people being confused? Are we sure we can fix the root cause of the confusion without changing the balance?
* If chaining is confusing, why have chaining at all? Currently our own build pipeline [looks like this](https://github.com/dagger/dagger/blob/main/magefile.go#L107-L122). If we think chaining confusing to developers, shouldn't we change our own pipeline to not use chaining at all? If not: why, what's the rule?

**Comment by helderco** [comment]
> @helderco You didn't say this explicitly, but I see you implemented the implicit approach -- I assume you prefer it as well?

I've actually said multiple times here that I prefer less verbose. I'm fine with the documented convention. If people want more explicitness, I like the simplicity of:
- Any leaf returns a `Leaf` object with an `execute()` method;
- Any object with an `id` field also includes an `execute()` method which uses that `id` internally (`id` would need to not be lazy);
- It only executes when you hit `execute()`. Until then, you're just building a query.

This requires little change, at least on my part.

> However I'm concerned about the Typescript SDK, since the execution needs to be `async`, we cannot really apply for an implicit pattern.

Why not? It should be the same as Python.

**Comment by sipsma** [comment]
Sort of late to the party here, but for Go I agree that we should wait for feedback from users after they start using the SDK w/ hidden IDs. Previously, any intuition users could develop about the model was shattered by `ID` being lazy yet returning error. I want to see whether intuition can develop now or if it's just shattered somewhere else (and if so, where).

---

If we do want more changes only thing I think what @vito hinted at here is worth expanding on:
> Maybe it's just a naming problem; Exec sounds more synchronous than WithX. I

We *could* make it so all our non-synchronous APIs are prefixed with `With`. So `WithExec` instead of `Exec`. This could also be combined with part of @shykes's suggestion: we could also change all synchronous APIs to be prefixed with `Get`.

So then the intuition becomes:
1. `With*` means lazy (I'm still just building up the object by including something *with* it)
2. `Get*` means "*get* me that value right now".

Part of this would imply that we change `.Stdout().Contents()` because `.WithStdout().GetContents()` doesn't make any sense. But tbh I already don't think it makes sense for those two to be split. `.GetStdout()` would make perfect sense to me though.
* Maybe it's a good thing that this sort of naming scheme can guide us like that. Or maybe it's way too restrictive. We'd need to sit down and go through the whole API to find any gotchas where it breaks and results in unintuitive names.

Not convinced by this at all, just throwing it out there, but the more general idea of baking the lazy vs. sync intuition in the naming scheme does seem nice in that it would probably work across any language too.

**Comment by slumbering** [comment]
> Node
> @slumbering mentioned implicit doesn't work in TypeScript

> However I'm concerned about the Typescript SDK, since the execution needs to be async, we cannot really apply for an implicit pattern.

Actually implicit does work, sorry if my explanation from yesterday was confusing. The problem was more how I was passing a container ID to another container while I was building a use case. The real issue is coming from how my Api works. 

```javascript
//  Run the Engine, execute the query implicitely and return the result 
const image = await new Api()
.container()
.from({address: "alpine"})
.exec({args: ["apk", "add", "yarn"]})
.fs()
.id()

// Re run an engine, so the image.id from above doesn't make sens anymore.
const sourceAfterInstall = await new Api()
  .container({id: image.id})
  .exec({args: ["yarn", "install"]})
...
  ```

**Comment by shykes** [comment]
> for Go I agree that we should wait for feedback from users after they start using the SDK w/ hidden IDs. Previously, any intuition users could develop about the model was shattered by `ID` being lazy yet returning error. I want to see whether intuition can develop now or if it's just shattered somewhere else (and if so, where).

I completely agree.

**Comment by helderco** [comment]
It's not resolved but since it's pending on user feedback we can reopen later if necessary.
